### PR TITLE
Improve instance reloading, introduce simple worker, add some convenience step functions

### LINKED ISSUE #275: allow activityId/taskDefinitionKey on "user_tasks_completes_with_variables"
Wen we execute a job, we can provide the activityId to ensure we are executing the correct expected job. We should provide this for user tasks as well so we can be sure that we do not accidentally complete st different.

### LINKED ISSUE #276: provide instance_is_finished_with_endEvent ...
To ensure my instance is ended and used a defined endEvent, I currently need two assertions. We should allow the ended assertion to verify the activityId of the end Event chosen

### LINKED ISSUE #278: Make sure the assertions on current instance are not using a stale instance but rather retrieve a new one from the runtime
The old process instance reference stored in the `Supplier<ProcessInstance>` is not ALWAYS delivering correct answers in terms of "waiting at", "finished" etc... Make sure, it is possible to reinitialize it with a fresh value.

### PR DESCRIPTION
**Pull request by zambrovski**
- adding new step functions
- fix #278 
- fix #276 
- fix #275

### CODE REVIEW

File: extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+open class SelfUpdatingInstanceSupplier(
+  protected val processInstanceId: String,
+  protected val runtimeService: RuntimeService
+) : Supplier<ProcessInstance> {
+
+  /**
+   * Retrieves the process instance.
+   */
+  override fun get(): ProcessInstance {
```
Comment (jangalinski) [inline extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt:19]: As discussed: Might be problematic as we wont get a processInstance once it is finished ... but there _could_ (not sure if and which) be BpmnAware assertions that still do things with the processInstance in Memory.

So a slightly better way would be to keep the last found instance and provide it the ended instance cannot be found anymore.

File: extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    val instances = runtimeService.createProcessInstanceQuery().active().processInstanceId(processInstanceId).list()
+    return when (instances.size) {
+      1 -> instances.first()
```
Comment (jangalinski) [inline extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt:22]: Isn't the size check just a fancy way of `.single() ?: throw ... `?
single() fails when more than 1 instance is found, but that is not possible when we query by id. Its 0 or 1.

**Comment by jangalinski** [comment]
Some remarks/questions ... otherwise, I do not see anything breaking so I would go for a release, and patch if required, can't think of any additional test cases right now.

File: extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/ProcessStage.kt [hunk 3]
```diff
@@ -303,6 +370,22 @@ class ProcessStage<SELF : ProcessStage<SELF, PROCESS_BEAN>, PROCESS_BEAN : Suppl
...
+   * Asserts that variable was set on a historic instance.
+   * @param variableName name of the variable.
+   * @param value value of the variable.
+   * @return fluent stage.
+   */
+  @As("variable \$variableName was set to \$value")
+  fun variable_was_set(@Quoted variableName: String, @Quoted value: Any): SELF = step {
+    val historicVariables = camunda.historyService.createHistoricVariableInstanceQuery().list()
+    assertThat(historicVariables.any { variable ->
+      variable.processInstanceId == processInstanceSupplier.get().processInstanceId
```
[the remark this answers is not in the document]
Comment (zambrovski) [reply extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/ProcessStage.kt:383]: agree

File: extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt (hunk 2 above)
Comment (zambrovski) [reply extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt:22]: agree

File: extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt (hunk 1 above)
Comment (zambrovski) [reply extension/src/main/kotlin/io/holunda/camunda/bpm/extension/jgiven/SelfUpdatingInstanceSupplier.kt:19]: yes

**Pull request merged**